---
title: "CellChat"
author: "heinin"
date: "2025-05-05"
output:
  workflowr::wflow_html:
    code_folding: hide
  html_notebook: default
editor_options:
  chunk_output_type: console
---

## Introduction

Ligand-receptor analysis using CellChat. The comparison is NEO vs. ADJ after CAR+anti-PD1 therapy.

## Packages and environment variables

```{r, warning=F, message=F}

suppressPackageStartupMessages({
  library(Seurat)
  library(SeuratObject)
  library(SeuratDisk)
  library(tidyverse)
  library(tibble)
  library(ggplot2)
  library(ggpubr)
  library(workflowr)
  library(googlesheets4)
  library(CellChat)})

setwd("/home/hnatri/SPP1_mouse_scRNAseq/")
set.seed(1234)
options(future.globals.maxSize = 30000 * 1024^2)
reduction <- "umap"

source("/home/hnatri/SPP1_mouse_scRNAseq/code/CART_plot_functions.R")
source("/home/hnatri/SPP1_mouse_scRNAseq/code/colors_themes.R")

```

## Run CellChat

CellChat was run for Day 16 only using code/CellChat.R.

## Inspecting results

### Import data

```{r, message = F, warning = F}

seurat_object <- readRDS("/tgen_labs/banovich/BCTCSF/SPP1_mouse_scRNAseq/scRNAseq_Seurat_dim8_granular_annot.rds")
#CC_objects_compare <- readRDS("/tgen_labs/banovich/BCTCSF/SPP1_mouse_scRNAseq/CART_SPP1_Kluc_Cellchat_compare.rsd")
#CC_merged_object <- readRDS("/tgen_labs/banovich/BCTCSF/SPP1_mouse_scRNAseq/CART_SPP1_Kluc_Cellchat_merged.rds")
CC_objects_compare <- readRDS("/scratch/hnatri/CART/CART_SPP1_Kluc_Cellchat_compare.rsd")
CC_merged_object <- readRDS("/scratch/hnatri/CART/CART_SPP1_Kluc_Cellchat_merged.rds")

# Sorting levels for plotting
#celltypes <- c(sort(c(grep("^M", unique(seurat_object$annot), value = T), c("Neut1"))),
#               sort(grep("^L", unique(seurat_object$annot), value = T)), c("NK", "Treg", "B1", "ILC", "DC"))
# annot_granular
celltypes <- sort(unique(seurat_object$annot_granular))

seurat_object$celltype <- seurat_object$annot_granular
seurat_object$celltype <- factor(seurat_object$celltype,
                                 levels = celltypes)

CC_objects_compare[[1]]@meta$celltype <- factor(CC_objects_compare[[1]]@meta$celltype,
                                                  levels = celltypes)
CC_objects_compare[[1]]@idents <- factor(CC_objects_compare[[1]]@idents,
                                         levels = celltypes)
CC_objects_compare[[2]]@meta$celltype <- factor(CC_objects_compare[[2]]@meta$celltype,
                                                  levels = celltypes)
CC_objects_compare[[2]]@idents <- factor(CC_objects_compare[[2]]@idents,
                                         levels = celltypes)

```

### UMAP of celltypes

```{r, message = F, warning = F, fig.width=5, fig.height=5}

DimPlot(seurat_object,
        group.by = "annot_granular",
        #cols = carspp1_celltype_col,
        reduction = "integratedSCTumap",
        label = T,
        label.box = T,
        label.size = 3,
        repel = T,
        raster = T,
        raster.dpi = c(1024, 1024),
        pt.size = 3) +
  ggtitle("") +
  theme_classic() +
  NoLegend() +
  NoAxes() +
  coord_fixed(1)

```

```{r, message = F, warning = F, fig.width=12, fig.height=6}

# Compare the total number of interactions and interaction strength
gg1 <- compareInteractions(CC_merged_object, show.legend = F, group = c(1,2))
gg2 <- compareInteractions(CC_merged_object, show.legend = F, group = c(1,2),
                           measure = "weight")
gg1 + gg2

```

```{r, message = F, warning = F, fig.width=12, fig.height=6}

# Differential number of interactions or interaction strength among different cell populations
par(mfrow = c(1,2), xpd=TRUE)
netVisual_diffInteraction(CC_merged_object, weight.scale = T)
netVisual_diffInteraction(CC_merged_object, weight.scale = T,
                          measure = "weight")

# Heatmaps
gg1 <- netVisual_heatmap(CC_merged_object,
                         cluster.rows = T,
                         cluster.cols = T)
gg2 <- netVisual_heatmap(CC_merged_object,
                         measure = "weight",
                         cluster.rows = T,
                         cluster.cols = T)
gg1 + gg2

# To better control the node size and edge weights of the inferred networks
# across different datasets, we compute the maximum number of cells per cell
# group and the maximum number of interactions (or interaction weights) across all datasets.
weight.max <- getMaxWeight(CC_objects_compare, attribute = c("idents","count"))
par(mfrow = c(1,2), xpd=TRUE)
for (i in 1:length(CC_objects_compare)) {
  netVisual_circle(CC_objects_compare[[i]]@net$count,
                   weight.scale = T,
                   label.edge= F,
                   edge.weight.max = weight.max[2],
                   edge.width.max = 12,
                   title.name = paste0("Number of interactions - ", names(CC_objects_compare)[i]),
                   color.use = carspp1_celltype_col)
}

# Need to run netAnalysis_computeCentrality before netAnalysis_signalingRole_scatter

# Compute the network centrality scores
CC_objects_compare <- lapply(CC_objects_compare, function(xx){
  netAnalysis_computeCentrality(xx, slot.name = "netP") # the slot 'netP' means the inferred intercellular communication network of signaling pathways
})

# Visualize the computed centrality scores using heatmap, allowing ready identification of major signaling roles of cell groups
#netAnalysis_signalingRole_network(CC_objects_compare[[1]], signaling = pathways.show, width = 8, height = 2.5, font.size = 10)

```

```{r, message = F, warning = F, fig.width=12, fig.height=6, eval = F}

# Compare the major sources and targets in 2D space
num.link <- sapply(CC_objects_compare, function(x) {rowSums(x@net$count) + colSums(x@net$count)-diag(x@net$count)})
weight.MinMax <- c(min(num.link), max(num.link)) # control the dot size in the different datasets
gg <- list()
for (i in 1:length(CC_objects_compare)) {
  gg[[i]] <- netAnalysis_signalingRole_scatter(CC_objects_compare[[i]],
                                               title = names(CC_objects_compare)[i],
                                               weight.MinMax = weight.MinMax,
                                               color.use = carspp1_celltype_col)
}

patchwork::wrap_plots(plots = gg)

```

## Identify and visualize the conserved and context-specific signaling pathways

```{r, fig.height=13}
# Identify and visualize the conserved and context-specific signaling pathways:
# signaling pathways, (i) turn off, (ii) decrease, (iii) turn on or (iv)
# increase, by change their information flow at one condition as compared to
# another condition.
# The top signaling pathways colored red are enriched in NEO, and these colored green were enriched in ADJ

# NEO vs. ADJ

# Extracting data from the plot
ranknet_res <- rankNet(CC_merged_object,
                       mode = "comparison",
                       stacked = F,
                       do.stat = T,
                       cutoff.pvalue = 0.01,
                       #tol = 1,
                       thresh = 0.01)

ranknet_res

```

```{r, message = F, warning = F}

ranknet_data <- ggplot_build(ranknet_res)
info_flow_data <- ranknet_data$plot$data

# Constructing the plot
info_flow_data$log2_contribution_1 <- log2(info_flow_data$contribution+1)
info_flow_data$group <- ifelse(info_flow_data$group == "CART_SPP1_CC", "CART_SPP1", "CART")

barplot_colors <- c("CART" = "deeppink3", "CART_SPP1_CC" = "aquamarine4")

hist(info_flow_data$contribution, breaks = 55, main = "", xlab = "Information flow")
abline(v=1, col = "red")

hist(info_flow_data$contribution, breaks = 20)

```

```{r, message = F, warning = F, fig.width=12}

# Plotting delta
barplot_colors <- c("aquamarine4", "deeppink3")

delta_plot <- info_flow_data %>%
  dplyr::group_by(name) %>%
  dplyr::filter(any(contribution>1)) %>%
  ungroup() %>%
  dplyr::group_by(group) %>%
  dplyr::filter(any(pvalues<0.01)) %>%
  ungroup() %>% 
  dplyr::select(name, contribution, group) %>%
  tidyr::pivot_wider(names_from = "group", values_from = "contribution") %>%
  mutate(delta = CART-CART_SPP1) %>%
  dplyr::select(name, delta) %>%
  #filter(abs(delta)>0.00002) %>%
  #mutate(log2_delta = log2(delta)) %>%
  mutate(sign = delta > 0) %>%
  ggplot(aes(x = reorder(name, -delta), y = delta, fill = sign)) +
  geom_bar(position = "dodge", stat="identity") +
  scale_fill_manual(values = barplot_colors, labels=c("CAR T + anti-SPP1", "CAR T")) +
  theme_bw() + 
  xlab("") + 
  ylab(expression(Delta ~ "information flow")) +
  labs(fill = "Treatment") +
  theme(axis.text.x = element_text(angle = 45, vjust = 0.9, hjust=0.9))

delta_plot

```

## M7 vs. others

```{r, message = F, warning = F, fig.width = 8, fig.height = 20}

# Compare all the interactions sending from M7 to L3 cells

#par(mfrow = c(1, 2), xpd = TRUE)
#netVisual_chord_gene(CC_objects_compare[["ADJ_CC"]], sources.use = c(18), targets.use = c(20), lab.cex = 0.5, title.name = paste0("Signaling from M7 - ADJ"))

#CC_merged_object$sources
levels(CC_merged_object@idents[[2]])

netVisual_bubble(CC_merged_object, sources.use = 10, 
                 #targets.use = setdiff(seq(1, 21), c(12)),
                 thresh = 0.05, comparison = c(1, 2), angle.x = 45)

```

## CD4 cells vs. others

```{r, message = F, warning = F, fig.width = 10, fig.height = 24}

levels(CC_merged_object@idents[[2]])

# Compare all the interactions sending from CD4
netVisual_bubble(CC_merged_object, sources.use = 3,
                 #targets.use = setdiff(seq(1, 21), c(16)),
                 thresh = 0.05, comparison = c(1, 2), angle.x = 45)

netVisual_bubble(CC_merged_object, sources.use = 14,
                 #targets.use = setdiff(seq(1, 21), c(3)),
                 thresh = 0.05, comparison = c(1, 2), angle.x = 45)

netVisual_bubble(CC_merged_object, sources.use = 21,
                 #targets.use = setdiff(seq(1, 21), c(13)),
                 thresh = 0.05, comparison = c(1, 2), angle.x = 45)

```

## Saving all significant pathways

```{r, message = F, warning = F, fig.width = 4, fig.height = 4}

# Using the netVisual_bubble function to pull all LR pair results
lr_all <- netVisual_bubble(CC_merged_object,
                           sources.use = NULL, targets.use = NULL,
                           comparison = c(1, 2), angle.x = 45)


lr_all_data <- ggplot_build(lr_all)
lr_all_data <- lr_all_data$plot$data

hist(lr_all_data$pval)

write.table(lr_all_data, "/home/hnatri/SPP1_mouse_scRNAseq/CARSPP1_vs_CAR_CellChat_LR_all_res.tsv",
            quote = F, sep = "\t", row.names = F)

```

## Selected pathways

```{r, message = F, warning = F, fig.width = 9, fig.height = 5}

pathways.show <- c("SPP1")
par(mfrow = c(1,2), xpd=TRUE)
ht <- list()
for (i in 1:length(CC_objects_compare)) {
  ht[[i]] <- netVisual_heatmap(CC_objects_compare[[i]],
                               signaling = pathways.show,
                               color.heatmap = "Reds",
                               title.name = paste(pathways.show, "signaling ", names(CC_objects_compare)[i]))
}

ComplexHeatmap::draw(ht[[1]] + ht[[2]], ht_gap = unit(0.5, "cm"))

pathways.show <- c("MHC-II")
par(mfrow = c(1,2), xpd=TRUE)
ht <- list()
for (i in 1:length(CC_objects_compare)) {
  ht[[i]] <- netVisual_heatmap(CC_objects_compare[[i]],
                               signaling = pathways.show,
                               color.heatmap = "Reds",
                               title.name = paste(pathways.show, "signaling ", names(CC_objects_compare)[i]))
}

ComplexHeatmap::draw(ht[[1]] + ht[[2]], ht_gap = unit(0.5, "cm"))

pathways.show <- c("MHC-I")
par(mfrow = c(1,2), xpd=TRUE)
ht <- list()
for (i in 1:length(CC_objects_compare)) {
  CC_objects_compare[[i]]@meta$celltype <- factor(CC_objects_compare[[i]]@meta$celltype,
                                                  levels = celltypes)
  CC_objects_compare[[i]]@idents <- factor(CC_objects_compare[[i]]@idents,
                                         levels = celltypes)

  ht[[i]] <- netVisual_heatmap(CC_objects_compare[[i]],
                               signaling = pathways.show,
                               color.heatmap = "Reds",
                               title.name = paste(pathways.show, "signaling ", names(CC_objects_compare)[i]))
}

ComplexHeatmap::draw(ht[[1]] + ht[[2]], ht_gap = unit(0.5, "cm"))

# ADGRE, ICAM, JAM and SEMA4
pathways.show <- c("ADGRE")
par(mfrow = c(1,2), xpd=TRUE)
ht <- list()
for (i in 1:length(CC_objects_compare)) {
  CC_objects_compare[[i]]@meta$celltype <- factor(CC_objects_compare[[i]]@meta$celltype,
                                                  levels = celltypes)
  CC_objects_compare[[i]]@idents <- factor(CC_objects_compare[[i]]@idents,
                                         levels = celltypes)

  ht[[i]] <- netVisual_heatmap(CC_objects_compare[[i]],
                               signaling = pathways.show,
                               color.heatmap = "Reds",
                               title.name = paste(pathways.show, "signaling ", names(CC_objects_compare)[i]))
}

ComplexHeatmap::draw(ht[[1]] + ht[[2]], ht_gap = unit(0.5, "cm"))

pathways.show <- c("ICAM")
par(mfrow = c(1,2), xpd=TRUE)
ht <- list()
for (i in 1:length(CC_objects_compare)) {
  CC_objects_compare[[i]]@meta$celltype <- factor(CC_objects_compare[[i]]@meta$celltype,
                                                  levels = celltypes)
  CC_objects_compare[[i]]@idents <- factor(CC_objects_compare[[i]]@idents,
                                         levels = celltypes)

  ht[[i]] <- netVisual_heatmap(CC_objects_compare[[i]],
                               signaling = pathways.show,
                               color.heatmap = "Reds",
                               title.name = paste(pathways.show, "signaling ", names(CC_objects_compare)[i]))
}

ComplexHeatmap::draw(ht[[1]] + ht[[2]], ht_gap = unit(0.5, "cm"))

pathways.show <- c("JAM")
par(mfrow = c(1,2), xpd=TRUE)
ht <- list()
for (i in 1:length(CC_objects_compare)) {
  CC_objects_compare[[i]]@meta$celltype <- factor(CC_objects_compare[[i]]@meta$celltype,
                                                  levels = celltypes)
  CC_objects_compare[[i]]@idents <- factor(CC_objects_compare[[i]]@idents,
                                         levels = celltypes)

  ht[[i]] <- netVisual_heatmap(CC_objects_compare[[i]],
                               signaling = pathways.show,
                               color.heatmap = "Reds",
                               title.name = paste(pathways.show, "signaling ", names(CC_objects_compare)[i]))
}

ComplexHeatmap::draw(ht[[1]] + ht[[2]], ht_gap = unit(0.5, "cm"))

pathways.show <- c("SEMA4")
par(mfrow = c(1,2), xpd=TRUE)
ht <- list()
for (i in 1:length(CC_objects_compare)) {
  CC_objects_compare[[i]]@meta$celltype <- factor(CC_objects_compare[[i]]@meta$celltype,
                                                  levels = celltypes)
  CC_objects_compare[[i]]@idents <- factor(CC_objects_compare[[i]]@idents,
                                         levels = celltypes)

  ht[[i]] <- netVisual_heatmap(CC_objects_compare[[i]],
                               signaling = pathways.show,
                               color.heatmap = "Reds",
                               title.name = paste(pathways.show, "signaling ", names(CC_objects_compare)[i]))
}

ComplexHeatmap::draw(ht[[1]] + ht[[2]], ht_gap = unit(0.5, "cm"))

# Chord diagram
pathways.show <- c("MHC-I") 
par(mfrow = c(1,2), xpd=TRUE)
for (i in 1:length(CC_objects_compare)) {
  netVisual_aggregate(CC_objects_compare[[i]],
                      signaling = pathways.show,
                      layout = "chord",
                      signaling.name = paste(pathways.show, names(CC_objects_compare)[i]),
                      color.use = carspp1_celltype_col)
}

pathways.show <- c("ApoE") 
par(mfrow = c(1,2), xpd=TRUE)
for (i in 1:length(CC_objects_compare)) {
  #group = group.cellType,
  netVisual_chord_cell(CC_objects_compare[[i]],
                       signaling = pathways.show,
                       title.name = paste0(pathways.show, " signaling network - ", names(CC_objects_compare)[i]),
                       color.use = carspp1_celltype_col)
}

par(mfrow = c(1, 2), xpd=TRUE)

# Compare the signaling gene expression distribution between different datasets
CC_merged_object@meta$Sample = factor(CC_merged_object@meta$Sample, levels = c("CART_SPP1", "CART")) # set factor level
plotGeneExpression(CC_merged_object,
                   signaling = "SPP1",
                   split.by = "Sample")
#color.use = c("red", "blue"))

DefaultAssay(seurat_object) <- "RNA"
seurat_object <- NormalizeData(seurat_object)
plot_features <- c("Spp1", "Cd44")
ccl_features <- rownames(seurat_object)[grep("Ccl", rownames(seurat_object))]

VlnPlot(seurat_object,
        features = plot_features,
        group.by = "annot",
        split.by = "orig.ident",
        pt.size = 0,
        slot = "data",
        ncol = 1) &
  theme_bw() &
  RotatedAxis() &
  xlab("") &
  RotatedAxis() 

VlnPlot(seurat_object,
        features = plot_features,
        group.by = "orig.ident",
        pt.size = 0,
        slot = "data",
        ncol = 1) &
  theme_bw() &
  RotatedAxis() &
  xlab("") &
  RotatedAxis()

VlnPlot(seurat_object,
        features = plot_features,
        group.by = "annot",
        split.by = "orig.ident",
        pt.size = 0,
        slot = "data",
        ncol = 1,
        cols = c("aquamarine3", "deeppink3")) &
  theme_bw() &
  xlab("") &
  ylab("Log-normalized expression") &
  RotatedAxis()
  #NoLegend()


```

### CD80/CD86 - CD28

```{r, message = F, warning = F, fig.width = 10, fig.height = 5}

pathways.show <- c("CD80_CD28")
par(mfrow = c(1,2), xpd=TRUE)
ht <- list()
for (i in 1:length(CC_objects_compare)) {
  CC_objects_compare[[i]]@meta$celltype <- factor(CC_objects_compare[[i]]@meta$celltype,
                                                  levels = celltypes)
  CC_objects_compare[[i]]@idents <- factor(CC_objects_compare[[i]]@idents,
                                         levels = celltypes)

  ht[[i]] <- netVisual_heatmap(CC_objects_compare[[i]],
                               signaling = pathways.show,
                               slot.name = "net",
                               color.heatmap = "Reds",
                               title.name = paste(pathways.show, "signaling ", names(CC_objects_compare)[i]))
}

ComplexHeatmap::draw(ht[[1]] + ht[[2]], ht_gap = unit(0.5, "cm"))

pathways.show <- c("CD86_CD28")
par(mfrow = c(1,2), xpd=TRUE)
ht <- list()
for (i in 1:length(CC_objects_compare)) {
  CC_objects_compare[[i]]@meta$celltype <- factor(CC_objects_compare[[i]]@meta$celltype,
                                                  levels = celltypes)
  CC_objects_compare[[i]]@idents <- factor(CC_objects_compare[[i]]@idents,
                                         levels = celltypes)

  ht[[i]] <- netVisual_heatmap(CC_objects_compare[[i]],
                               signaling = pathways.show,
                               slot.name = "net",
                               color.heatmap = "Reds",
                               title.name = paste(pathways.show, "signaling ", names(CC_objects_compare)[i]))
}

ComplexHeatmap::draw(ht[[1]] + ht[[2]], ht_gap = unit(0.5, "cm"))

```


## Delta signaling heatmaps

### SPP1

NAs in each signaling matrix due to lack of signaling

```{r, message = F, warning = F, fig.width = 6, fig.height = 5}

# Delta heatmap
pathways.show <- c("SPP1")
par(mfrow = c(1,2), xpd=TRUE)
ht <- list("CART_SPP1_CC" = NULL,
           "CART_CC" = NULL)
for (i in names(CC_objects_compare)) {
  CC_objects_compare[[i]]@meta$annot_granular <- factor(CC_objects_compare[[i]]@meta$annot_granular,
                                                  levels = sort(unique(CC_objects_compare[[i]]@meta$annot_granular)))
  CC_objects_compare[[i]]@idents <- factor(CC_objects_compare[[i]]@idents,
                                           levels = sort(unique(CC_objects_compare[[i]]@meta$annot_granular)))
  
  ht[[i]] <- netVisual_heatmap(object = CC_objects_compare[[i]],
                               signaling = pathways.show,
                               color.heatmap = c('#ffffff','#b2182b'),
                               measure = "count",
                               slot.name = "netP",
                               #remove.isolate = T,
                               #cluster.rows = T,
                               #cluster.cols = T,
                               title.name = paste(pathways.show, "signaling ", names(CC_objects_compare)[i]))
                               #color.use = immune_fibro_celltype_col)
}


delta_ht <- ht[[2]]
delta_ht@matrix <- ht[[2]]@matrix - ht[[1]]@matrix

delta_ht@matrix_color_mapping

mx_cols <- c("#4287F5FF", "#ffffff", "#A84E76FF", "#B12A3DFF", "#B2182BFF")
mx_breaks <- c(-0.1, 0, 0.1, 0.2, 0.3)
col_fun = colorRamp2(mx_breaks, mx_cols)

delta_ht@matrix_color_mapping <- ComplexHeatmap::ColorMapping(col_fun = col_fun,
                                                              breaks = mx_breaks,
                                                              na_col = "#e6e7e8")

ComplexHeatmap::draw(delta_ht)

#filename <- "CellChat_SPP1_delta_heatmap.pdf"
#pdf(file = filename,
#    width = 4,
#    height = 3.5)
#
#ComplexHeatmap::draw(delta_ht)
#
#dev.off()

```

### MHC-I

```{r, message = F, warning = F, fig.width = 6, fig.height = 5}

# Delta heatmap
pathways.show <- c("MHC-I")
par(mfrow = c(1,2), xpd=TRUE)
ht <- list("CART_SPP1_CC" = NULL,
           "CART_CC" = NULL)
for (i in names(CC_objects_compare)) {
  CC_objects_compare[[i]]@meta$annot_granular <- factor(CC_objects_compare[[i]]@meta$annot_granular,
                                                  levels = sort(unique(CC_objects_compare[[i]]@meta$annot_granular)))
  CC_objects_compare[[i]]@idents <- factor(CC_objects_compare[[i]]@idents,
                                           levels = sort(unique(CC_objects_compare[[i]]@meta$annot_granular)))
  
  ht[[i]] <- netVisual_heatmap(object = CC_objects_compare[[i]],
                               signaling = pathways.show,
                               color.heatmap = c('#ffffff','#b2182b'),
                               measure = "count",
                               slot.name = "netP",
                               #remove.isolate = T,
                               #cluster.rows = T,
                               #cluster.cols = T,
                               title.name = paste(pathways.show, "signaling ", names(CC_objects_compare)[i]))
                               #color.use = immune_fibro_celltype_col)
}


delta_ht <- ht[[2]]
delta_ht@matrix <- ht[[2]]@matrix - ht[[1]]@matrix

delta_ht@matrix_color_mapping

mx_cols <- c("#4287F5FF", "#ffffff", "#A84E76FF", "#B12A3DFF", "#B2182BFF")
mx_breaks <- c(0, 0.2, 0.4, 0.6, 0.8)
col_fun = colorRamp2(mx_breaks, mx_cols)

delta_ht@matrix_color_mapping <- ComplexHeatmap::ColorMapping(col_fun = col_fun,
                                                              breaks = mx_breaks,
                                                              na_col = "#e6e7e8")

ComplexHeatmap::draw(delta_ht)

```

### MHC-II

```{r, message = F, warning = F, fig.width = 6, fig.height = 5}

# Delta heatmap
pathways.show <- c("MHC-II")
par(mfrow = c(1,2), xpd=TRUE)
ht <- list("CART_SPP1_CC" = NULL,
           "CART_CC" = NULL)
for (i in names(CC_objects_compare)) {
  CC_objects_compare[[i]]@meta$annot_granular <- factor(CC_objects_compare[[i]]@meta$annot_granular,
                                                  levels = sort(unique(CC_objects_compare[[i]]@meta$annot_granular)))
  CC_objects_compare[[i]]@idents <- factor(CC_objects_compare[[i]]@idents,
                                           levels = sort(unique(CC_objects_compare[[i]]@meta$annot_granular)))
  
  ht[[i]] <- netVisual_heatmap(object = CC_objects_compare[[i]],
                               signaling = pathways.show,
                               color.heatmap = c('#ffffff','#b2182b'),
                               measure = "count",
                               slot.name = "netP",
                               #remove.isolate = T,
                               #cluster.rows = T,
                               #cluster.cols = T,
                               title.name = paste(pathways.show, "signaling ", names(CC_objects_compare)[i]))
                               #color.use = immune_fibro_celltype_col)
}


delta_ht <- ht[[2]]
delta_ht@matrix <- ht[[2]]@matrix - ht[[1]]@matrix

delta_ht@matrix_color_mapping

mx_cols <- c("#ffffff", "#A84E76FF", "#B12A3DFF", "#B2182BFF")
mx_breaks <- c(0, 0.5, 1.0, 1.5)
col_fun = colorRamp2(mx_breaks, mx_cols)

delta_ht@matrix_color_mapping <- ComplexHeatmap::ColorMapping(col_fun = col_fun,
                                                              breaks = mx_breaks,
                                                              na_col = "#e6e7e8")

ComplexHeatmap::draw(delta_ht)

```

## Delta heatmaps with NA = 0

### SPP1

```{r, message = F, warning = F, fig.width = 6, fig.height = 5}

CC_objects_compare <- readRDS("/scratch/hnatri/CART/CART_SPP1_Kluc_Cellchat_compare_p1.rsd")

# Delta heatmap
pathways.show <- c("SPP1")
par(mfrow = c(1,2), xpd=TRUE)
ht <- list("CART_SPP1_CC" = NULL,
           "CART_CC" = NULL)
for (i in names(CC_objects_compare)) {
  CC_objects_compare[[i]]@meta$annot_granular <- factor(CC_objects_compare[[i]]@meta$annot_granular,
                                                  levels = sort(unique(CC_objects_compare[[i]]@meta$annot_granular)))
  CC_objects_compare[[i]]@idents <- factor(CC_objects_compare[[i]]@idents,
                                           levels = sort(unique(CC_objects_compare[[i]]@meta$annot_granular)))
  
  ht[[i]] <- netVisual_heatmap(object = CC_objects_compare[[i]],
                               signaling = pathways.show,
                               color.heatmap = c('#ffffff','#b2182b'),
                               measure = "count",
                               slot.name = "netP",
                               #remove.isolate = T,
                               #cluster.rows = T,
                               #cluster.cols = T,
                               title.name = paste(pathways.show, "signaling"))
                               #color.use = immune_fibro_celltype_col)
}


delta_ht <- ht[[2]]

ht[[2]]@matrix[is.na(ht[[2]]@matrix)] <- 0
ht[[1]]@matrix[is.na(ht[[1]]@matrix)] <- 0

delta_ht@matrix <- ht[[2]]@matrix - ht[[1]]@matrix

delta_ht@matrix_color_mapping

mx_cols <- c("#ffffff", "#A84E76FF", "#B12A3DFF", "#B2182BFF")
mx_breaks <- c(0, 0.1, 0.2, 0.3)
col_fun = colorRamp2(mx_breaks, mx_cols)

delta_ht@matrix_color_mapping <- ComplexHeatmap::ColorMapping(col_fun = col_fun,
                                                              breaks = mx_breaks,
                                                              na_col = "#e6e7e8")

ComplexHeatmap::draw(delta_ht)

```

### MHC-I

```{r, message = F, warning = F, fig.width = 6, fig.height = 5}

# Delta heatmap
pathways.show <- c("MHC-I")
par(mfrow = c(1,2), xpd=TRUE)
ht <- list("CART_SPP1_CC" = NULL,
           "CART_CC" = NULL)
for (i in names(CC_objects_compare)) {
  CC_objects_compare[[i]]@meta$annot_granular <- factor(CC_objects_compare[[i]]@meta$annot_granular,
                                                  levels = sort(unique(CC_objects_compare[[i]]@meta$annot_granular)))
  CC_objects_compare[[i]]@idents <- factor(CC_objects_compare[[i]]@idents,
                                           levels = sort(unique(CC_objects_compare[[i]]@meta$annot_granular)))
  
  ht[[i]] <- netVisual_heatmap(object = CC_objects_compare[[i]],
                               signaling = pathways.show,
                               color.heatmap = c('#ffffff','#b2182b'),
                               measure = "count",
                               slot.name = "netP",
                               #remove.isolate = T,
                               #cluster.rows = T,
                               #cluster.cols = T,
                               title.name = paste(pathways.show, "signaling"))
                               #color.use = immune_fibro_celltype_col)
}


delta_ht <- ht[[2]]

ht[[2]]@matrix[is.na(ht[[2]]@matrix)] <- 0
ht[[1]]@matrix[is.na(ht[[1]]@matrix)] <- 0

delta_ht@matrix <- ht[[2]]@matrix - ht[[1]]@matrix

delta_ht@matrix_color_mapping

mx_cols <- c("#ffffff", "#A84E76FF", "#B12A3DFF", "#B2182BFF")
mx_breaks <- c(0, 0.5, 1.0, 1.5)
col_fun = colorRamp2(mx_breaks, mx_cols)

delta_ht@matrix_color_mapping <- ComplexHeatmap::ColorMapping(col_fun = col_fun,
                                                              breaks = mx_breaks,
                                                              na_col = "#e6e7e8")

ComplexHeatmap::draw(delta_ht)

```

### MHC-II

```{r, message = F, warning = F, fig.width = 6, fig.height = 5}

# Delta heatmap
pathways.show <- c("MHC-II")
par(mfrow = c(1,2), xpd=TRUE)
ht <- list("CART_SPP1_CC" = NULL,
           "CART_CC" = NULL)
for (i in names(CC_objects_compare)) {
  CC_objects_compare[[i]]@meta$annot_granular <- factor(CC_objects_compare[[i]]@meta$annot_granular,
                                                  levels = sort(unique(CC_objects_compare[[i]]@meta$annot_granular)))
  CC_objects_compare[[i]]@idents <- factor(CC_objects_compare[[i]]@idents,
                                           levels = sort(unique(CC_objects_compare[[i]]@meta$annot_granular)))
  
  ht[[i]] <- netVisual_heatmap(object = CC_objects_compare[[i]],
                               signaling = pathways.show,
                               color.heatmap = c('#ffffff','#b2182b'),
                               measure = "count",
                               slot.name = "netP",
                               #remove.isolate = T,
                               #cluster.rows = T,
                               #cluster.cols = T,
                               title.name = paste(pathways.show, "signaling"))
                               #color.use = immune_fibro_celltype_col)
}


delta_ht <- ht[[2]]

ht[[2]]@matrix[is.na(ht[[2]]@matrix)] <- 0
ht[[1]]@matrix[is.na(ht[[1]]@matrix)] <- 0

delta_ht@matrix <- ht[[2]]@matrix - ht[[1]]@matrix

delta_ht@matrix_color_mapping

mx_cols <- c("#ffffff", "#A84E76FF", "#B12A3DFF", "#B2182BFF")
mx_breaks <- c(0, 0.5, 1.0, 1.5)
col_fun = colorRamp2(mx_breaks, mx_cols)

delta_ht@matrix_color_mapping <- ComplexHeatmap::ColorMapping(col_fun = col_fun,
                                                              breaks = mx_breaks,
                                                              na_col = "#e6e7e8")

ComplexHeatmap::draw(delta_ht)

```

### ADGRE

```{r, message = F, warning = F, fig.width = 6, fig.height = 5}

# Delta heatmap
pathways.show <- c("ADGRE")
par(mfrow = c(1,2), xpd=TRUE)
ht <- list("CART_SPP1_CC" = NULL,
           "CART_CC" = NULL)
for (i in names(CC_objects_compare)) {
  CC_objects_compare[[i]]@meta$annot_granular <- factor(CC_objects_compare[[i]]@meta$annot_granular,
                                                  levels = sort(unique(CC_objects_compare[[i]]@meta$annot_granular)))
  CC_objects_compare[[i]]@idents <- factor(CC_objects_compare[[i]]@idents,
                                           levels = sort(unique(CC_objects_compare[[i]]@meta$annot_granular)))
  
  ht[[i]] <- netVisual_heatmap(object = CC_objects_compare[[i]],
                               signaling = pathways.show,
                               color.heatmap = c('#ffffff','#b2182b'),
                               measure = "count",
                               slot.name = "netP",
                               #remove.isolate = T,
                               #cluster.rows = T,
                               #cluster.cols = T,
                               title.name = paste(pathways.show, "signaling"))
                               #color.use = immune_fibro_celltype_col)
}


delta_ht <- ht[[2]]

ht[[2]]@matrix[is.na(ht[[2]]@matrix)] <- 0
ht[[1]]@matrix[is.na(ht[[1]]@matrix)] <- 0

delta_ht@matrix <- ht[[2]]@matrix - ht[[1]]@matrix

delta_ht@matrix_color_mapping

mx_cols <- c("#ffffff", "#A84E76FF", "#B12A3DFF", "#B2182BFF")
mx_breaks <- c(0, 0.5, 1.0, 1.5)
col_fun = colorRamp2(mx_breaks, mx_cols)

delta_ht@matrix_color_mapping <- ComplexHeatmap::ColorMapping(col_fun = col_fun,
                                                              breaks = mx_breaks,
                                                              na_col = "#e6e7e8")

ComplexHeatmap::draw(delta_ht)

```

### ICAM

```{r, message = F, warning = F, fig.width = 6, fig.height = 5}

# Delta heatmap
pathways.show <- c("ICAM")
par(mfrow = c(1,2), xpd=TRUE)
ht <- list("CART_SPP1_CC" = NULL,
           "CART_CC" = NULL)
for (i in names(CC_objects_compare)) {
  CC_objects_compare[[i]]@meta$annot_granular <- factor(CC_objects_compare[[i]]@meta$annot_granular,
                                                  levels = sort(unique(CC_objects_compare[[i]]@meta$annot_granular)))
  CC_objects_compare[[i]]@idents <- factor(CC_objects_compare[[i]]@idents,
                                           levels = sort(unique(CC_objects_compare[[i]]@meta$annot_granular)))
  
  ht[[i]] <- netVisual_heatmap(object = CC_objects_compare[[i]],
                               signaling = pathways.show,
                               color.heatmap = c('#ffffff','#b2182b'),
                               measure = "count",
                               slot.name = "netP",
                               #remove.isolate = T,
                               #cluster.rows = T,
                               #cluster.cols = T,
                               title.name = paste(pathways.show, "signaling"))
                               #color.use = immune_fibro_celltype_col)
}


delta_ht <- ht[[2]]

ht[[2]]@matrix[is.na(ht[[2]]@matrix)] <- 0
ht[[1]]@matrix[is.na(ht[[1]]@matrix)] <- 0

delta_ht@matrix <- ht[[2]]@matrix - ht[[1]]@matrix

delta_ht@matrix_color_mapping

mx_cols <- c("#ffffff", "#A84E76FF", "#B12A3DFF", "#B2182BFF")
mx_breaks <- c(0, 0.5, 1.0, 1.5)
col_fun = colorRamp2(mx_breaks, mx_cols)

delta_ht@matrix_color_mapping <- ComplexHeatmap::ColorMapping(col_fun = col_fun,
                                                              breaks = mx_breaks,
                                                              na_col = "#e6e7e8")

ComplexHeatmap::draw(delta_ht)

```

### JAM

```{r, message = F, warning = F, fig.width = 6, fig.height = 5}

# Delta heatmap
pathways.show <- c("JAM")
par(mfrow = c(1,2), xpd=TRUE)
ht <- list("CART_SPP1_CC" = NULL,
           "CART_CC" = NULL)
for (i in names(CC_objects_compare)) {
  CC_objects_compare[[i]]@meta$annot_granular <- factor(CC_objects_compare[[i]]@meta$annot_granular,
                                                  levels = sort(unique(CC_objects_compare[[i]]@meta$annot_granular)))
  CC_objects_compare[[i]]@idents <- factor(CC_objects_compare[[i]]@idents,
                                           levels = sort(unique(CC_objects_compare[[i]]@meta$annot_granular)))
  
  ht[[i]] <- netVisual_heatmap(object = CC_objects_compare[[i]],
                               signaling = pathways.show,
                               color.heatmap = c('#ffffff','#b2182b'),
                               measure = "count",
                               slot.name = "netP",
                               #remove.isolate = T,
                               #cluster.rows = T,
                               #cluster.cols = T,
                               title.name = paste(pathways.show, "signaling"))
                               #color.use = immune_fibro_celltype_col)
}


delta_ht <- ht[[2]]

ht[[2]]@matrix[is.na(ht[[2]]@matrix)] <- 0
ht[[1]]@matrix[is.na(ht[[1]]@matrix)] <- 0

delta_ht@matrix <- ht[[2]]@matrix - ht[[1]]@matrix

delta_ht@matrix_color_mapping

mx_cols <- c("#ffffff", "#A84E76FF", "#B12A3DFF", "#B2182BFF")
mx_breaks <- c(0, 0.5, 1.0, 1.5)
col_fun = colorRamp2(mx_breaks, mx_cols)

delta_ht@matrix_color_mapping <- ComplexHeatmap::ColorMapping(col_fun = col_fun,
                                                              breaks = mx_breaks,
                                                              na_col = "#e6e7e8")

ComplexHeatmap::draw(delta_ht)

```

### CD80 / CD28

```{r, message = F, warning = F, fig.width = 6, fig.height = 5}

# Delta heatmap
pathways.show <- c("CD80_CD28")
par(mfrow = c(1,2), xpd=TRUE)
ht <- list("CART_SPP1_CC" = NULL,
           "CART_CC" = NULL)
for (i in names(CC_objects_compare)) {
  CC_objects_compare[[i]]@meta$annot_granular <- factor(CC_objects_compare[[i]]@meta$annot_granular,
                                                  levels = sort(unique(CC_objects_compare[[i]]@meta$annot_granular)))
  CC_objects_compare[[i]]@idents <- factor(CC_objects_compare[[i]]@idents,
                                           levels = sort(unique(CC_objects_compare[[i]]@meta$annot_granular)))
  
  ht[[i]] <- netVisual_heatmap(object = CC_objects_compare[[i]],
                               signaling = pathways.show,
                               color.heatmap = c('#ffffff','#b2182b'),
                               measure = "count",
                               slot.name = "net",
                               #remove.isolate = T,
                               #cluster.rows = T,
                               #cluster.cols = T,
                               title.name = paste(pathways.show, "signaling"))
                               #color.use = immune_fibro_celltype_col)
}


delta_ht <- ht[[2]]

ht[[2]]@matrix[is.na(ht[[2]]@matrix)] <- 0
ht[[1]]@matrix[is.na(ht[[1]]@matrix)] <- 0

delta_ht@matrix <- ht[[2]]@matrix - ht[[1]]@matrix

delta_ht@matrix_color_mapping

mx_cols <- c("#ffffff", "#A84E76FF", "#B12A3DFF", "#B2182BFF")
mx_breaks <- c(0, 0.5, 1.0, 1.5)
col_fun = colorRamp2(mx_breaks, mx_cols)

delta_ht@matrix_color_mapping <- ComplexHeatmap::ColorMapping(col_fun = col_fun,
                                                              breaks = mx_breaks,
                                                              na_col = "#e6e7e8")

ComplexHeatmap::draw(delta_ht)

```

### CD86 / CD28

```{r, message = F, warning = F, fig.width = 6, fig.height = 5}

# Delta heatmap
pathways.show <- c("CD86_CD28")
par(mfrow = c(1,2), xpd=TRUE)
ht <- list("CART_SPP1_CC" = NULL,
           "CART_CC" = NULL)
for (i in names(CC_objects_compare)) {
  CC_objects_compare[[i]]@meta$annot_granular <- factor(CC_objects_compare[[i]]@meta$annot_granular,
                                                  levels = sort(unique(CC_objects_compare[[i]]@meta$annot_granular)))
  CC_objects_compare[[i]]@idents <- factor(CC_objects_compare[[i]]@idents,
                                           levels = sort(unique(CC_objects_compare[[i]]@meta$annot_granular)))
  
  ht[[i]] <- netVisual_heatmap(object = CC_objects_compare[[i]],
                               signaling = pathways.show,
                               color.heatmap = c('#ffffff','#b2182b'),
                               measure = "count",
                               slot.name = "net",
                               #remove.isolate = T,
                               #cluster.rows = T,
                               #cluster.cols = T,
                               title.name = paste(pathways.show, "signaling"))
                               #color.use = immune_fibro_celltype_col)
}


delta_ht <- ht[[2]]

ht[[2]]@matrix[is.na(ht[[2]]@matrix)] <- 0
ht[[1]]@matrix[is.na(ht[[1]]@matrix)] <- 0

delta_ht@matrix <- ht[[2]]@matrix - ht[[1]]@matrix

delta_ht@matrix_color_mapping

mx_cols <- c("#ffffff", "#A84E76FF", "#B12A3DFF", "#B2182BFF")
mx_breaks <- c(0, 0.5, 1.0, 1.5)
col_fun = colorRamp2(mx_breaks, mx_cols)

delta_ht@matrix_color_mapping <- ComplexHeatmap::ColorMapping(col_fun = col_fun,
                                                              breaks = mx_breaks,
                                                              na_col = "#e6e7e8")

ComplexHeatmap::draw(delta_ht)

# To build on command line, run Rscript -e "rmarkdown::render('CellChat.Rmd')"
# Then "mv *.html /home/hnatri/SPP1_mouse_scRNAseq/docs/"

```

