---
title: "CAR+anti-SPP1 Kluc tumor comparative analysis"
author: "heinin"
date: "2025-04-10"
output:
  workflowr::wflow_html:
    code_folding: hide
  html_notebook: default
editor_options:
  chunk_output_type: console
---

## Libraries and environment variables

```{r, message = F, warning = F, results = 'hide'}

library(workflowr)
library(Seurat)
library(googlesheets4)
library(tidyverse)
library(plyr)
library(ggrepel)
library(patchwork)
library(scProportionTest)

setwd("/home/hnatri/SPP1_mouse_scRNAseq/")
set.seed(1234)
options(future.globals.maxSize = 30000 * 1024^2)
reduction <- "integratedSCTumap"

source("/home/hnatri/SPP1_mouse_scRNAseq/code/CART_plot_functions.R")
source("/home/hnatri/SPP1_mouse_scRNAseq/code/colors_themes.R")

# Cluster annotations
#gs4_deauth()
#cluster_annot  <- gs4_get("https://docs.google.com/spreadsheets/d/127J6C4KF7uBGKUnrPuC1mcsb_wNCN6k1zXKSCbJ6q0M/edit?us#p=sharing")
#cluster_annot <- read_sheet(cluster_annot, sheet = "Cluster annotation")

```

## Import data

Annotations were updated after subclustering and annotating myeloid and lymphoid cells.

```{r, message = F, warning = F, results = 'hide', fig.width = 6, fig.height = 6}

seurat_data <- readRDS("/tgen_labs/banovich/BCTCSF/SPP1_mouse_scRNAseq/scRNAseq_Seurat_dim8_nCount1k_nFeature500_dblrate15_scGSEAmodules_reclustered_refAnnot.rds")

# Updating annotations
#seurat_data$annot <- mapvalues(x = seurat_data$sub.cluster,
#                               from = cluster_annot$sub.cluster,
#                               to = cluster_annot$annot)

DimPlot(seurat_data,
        group.by = "annot_granular",
        #cols = carspp1_celltype_col,
        reduction = reduction,
        label = T,
        label.box = T,
        label.size = 3,
        repel = T,
        raster = T,
        raster.dpi = c(1024, 1024),
        pt.size = 3) +
  ggtitle("") +
  theme_classic() +
  NoLegend() +
  NoAxes() +
  coord_fixed(1)

seurat_data$Sample <- seurat_data$orig.ident
seurat_data$Sample <- gsub("SPP1\\+CAR", "CAR_SPP1", seurat_data$Sample)
seurat_data$Sample <- gsub("TUMOR", "CTRL", seurat_data$Sample)

# Celltype colors
carspp1_celltype_col <- colorRampPalette(brewer.pal(10, "Paired"))(nb.cols <- length(unique(seurat_data$annot_granular)))
names(carspp1_celltype_col) <- sort(unique(seurat_data$annot_granular))

DimPlot(seurat_data,
        group.by = "annot_granular",
        reduction = "integratedSCTumap",
        cols = carspp1_celltype_col,
        label = T,
        repel = T,
        label.box = T) +
  coord_fixed() +
  NoLegend()

```

UMAPs of cell types by group

```{r, message = F, warning = F, fig.width = 10, fig.height = 4, results = 'hide', eval = F}

DimPlot(seurat_data,
        group.by = "annot_granular",
        split.by = "Sample",
        reduction = "integratedSCTumap",
        cols = carspp1_celltype_col) +
        #label = T,
        #repel = T,
        #label.box = T) +
  coord_fixed()

```

## Cell type proportions

```{r, message = F, warning = F, results = 'hide', fig.width = 8, fig.height = 5, eval = T}

create_barplot(seurat_data,
               group_var = "Sample",
               plot_var = "annot_granular",
               plot_levels = sort((unique(seurat_data$annot_granular))),
               group_levels = c("CTRL", "SPP1", "CAR", "CAR_SPP1"),
               plot_colors = carspp1_celltype_col,
               var_names =  c("Frequency (%)", ""),
               legend_title = "Celltype")

```

Using scProportionTest. The first group (for example, CAR in "CAR vs. CTRL") gets positive values.

```{r, message = F, warning = F, fig.width = 4, fig.height = 4, results = 'hide', eval = F}

prop_test <- sc_utils(seurat_data)

# Permutation testing and bootstrapping

# CAR vs. CTRL
prop_test <- permutation_test(
  prop_test, cluster_identity = "annot_granular",
  sample_2 = "CAR", sample_1 = "TUMOR",
  sample_identity = "orig.ident")

perm_plot <- permutation_plot(prop_test)

perm_plot + scale_colour_manual(values = c("tomato", "azure2")) + NoLegend() + ggtitle("CAR vs. CTRL")

# SPP1 vs. CTRL
prop_test <- permutation_test(
  prop_test, cluster_identity = "annot_granular",
  sample_2 = "SPP1", sample_1 = "TUMOR",
  sample_identity = "orig.ident")

perm_plot <- permutation_plot(prop_test)

perm_plot + scale_colour_manual(values = c("tomato", "azure2")) + NoLegend() + ggtitle("SPP1 vs. CTRL")

# CAR+SPP1 vs. CTRL
prop_test <- permutation_test(
  prop_test, cluster_identity = "annot_granular",
  sample_2 = "SPP1+CAR", sample_1 = "TUMOR",
  sample_identity = "orig.ident")

perm_plot <- permutation_plot(prop_test)

perm_plot + scale_colour_manual(values = c("tomato", "azure2")) + NoLegend() + ggtitle("SPP1+CAR vs. CTRL")

# CAR+SPP1 vs. CAR
prop_test <- permutation_test(
  prop_test, cluster_identity = "annot_granular",
  sample_2 = "SPP1+CAR", sample_1 = "CAR",
  sample_identity = "orig.ident")

perm_plot <- permutation_plot(prop_test)

perm_plot + scale_colour_manual(values = c("tomato", "azure2")) + NoLegend() + ggtitle("SPP1+CAR vs. CAR")

# CAR+SPP1 vs. SPP1
prop_test <- permutation_test(
  prop_test, cluster_identity = "annot_granular",
  sample_2 = "SPP1+CAR", sample_1 = "SPP1",
  sample_identity = "orig.ident")

perm_plot <- permutation_plot(prop_test)

perm_plot + scale_colour_manual(values = c("tomato", "azure2")) + NoLegend() + ggtitle("SPP1+CAR vs. SPP1")

```

## Comparing gene expression levels

## Mouse modules

```{r, message = F, warning = F, fig.width = 10, fig.height = 10, results = 'hide'}

VlnPlot(seurat_data,
        features =  c("DAM", "Homeostatic", "MHC_I_machinery", "MHC_II_machinery", "Costimulation_APCB", "Adhesion_immunesynapse"),
        group.by = "orig.ident",
        ncol = 2,
        pt.size = 0)

```

### ANOVA and Tukey HSD

DAM

```{r, message = F, warning = F, fig.width = 12, fig.height = 8}

model <- aov(DAM ~ orig.ident, data = seurat_data@meta.data)
summary(model)

TukeyHSD(model)

```

Homeostatic

```{r, message = F, warning = F, fig.width = 12, fig.height = 8}

model <- aov(Homeostatic ~ orig.ident, data = seurat_data@meta.data)
summary(model)

TukeyHSD(model)

```

MHC_I_machinery

```{r, message = F, warning = F, fig.width = 12, fig.height = 8}

model <- aov(MHC_I_machinery ~ orig.ident, data = seurat_data@meta.data)
summary(model)

TukeyHSD(model)

```

MHC_II_machinery

```{r, message = F, warning = F, fig.width = 12, fig.height = 8}

model <- aov(MHC_II_machinery ~ orig.ident, data = seurat_data@meta.data)
summary(model)

TukeyHSD(model)

```

Costimulation_APCB

```{r, message = F, warning = F, fig.width = 10, fig.height = 6}

model <- aov(Costimulation_APCB ~ orig.ident, data = seurat_data@meta.data)
summary(model)

TukeyHSD(model)

```

Adhesion_immunesynapse

```{r, message = F, warning = F, fig.width = 12, fig.height = 8}

model <- aov(Adhesion_immunesynapse ~ orig.ident, data = seurat_data@meta.data)
summary(model)

TukeyHSD(model)

```

```{r, message = F, warning = F, fig.width = 14, fig.height = 10, results = 'hide', eval = F}

VlnPlot(seurat_data,
        features =  c("DAM", "Homeostatic", "MHC_I_machinery", "MHC_II_machinery", "Costimulation_APCB", "Adhesion_immunesynapse"),
        group.by = "snn_res.0.8",
        split.by = "orig.ident",
        ncol = 1,
        pt.size = 0)

```

## DEGs by cell population

### Myeloid

```{r, message = F, warning = F, results = 'hide', fig.width = 4, fig.height = 4}

# Rscript -e "rmarkdown::render('comparative_analysis.Rmd')"
# Then "mv *.html /home/hnatri/SPP1_mouse_scRNAseq/docs/"


```

### Lymphoid

